src/libinjector/linux/methods/linux_execve.c

```diff
@@ -0,0 +1,292 @@
+#include <libinjector/debug_helpers.h>
+
+#include "linux_execve.h"
+#include "linux_syscalls.h"
+
+static event_response_t cleanup(injector_t injector, x86_registers_t* regs);
+bool is_child_process(injector_t injector,  drakvuf_trap_info_t* info)
+{
+    if (info->proc_data.ppid == injector->target_pid)
+    {
+        PRINT_DEBUG("Inside child process\n");
+        return true;
+    }
+    PRINT_DEBUG("Inside parent process: %d\n", info->proc_data.pid);
+    PRINT_DEBUG("Step of injector: %d\n", injector->step + 1);
+    return false;
+}
+
+/* This function handles execve syscalls, it does so in a total of 5 steps
+ *
+ * STEP1:
+ * It initialises the syscalls and then it calls mmap to reserve some space for keeping
+ * the string arguments to be passed in execve
+ *
+ * STEP2:
+ * This step vforks the process and loosens the check around check_userspace_int3_trap
+ * but setting injector->fork = true. This helps us get the pid of the child process
+ * as soon as we get it, we use that child pid to tighten the checks again
+ *
+ * STEP3:
+ * We get the pid in this as this trap should only hit in the child process as per the
+ * checks in check_userspace_int3_trap. So we can now store the pid of the child process
+ * and setup execve calls
+ *
+ * STEP4:
+ * This step is considered a part of the cleanup, It checks if execve succeeded or not,
+ * if it did not then it will exit the child process and then keep the step so that we
+ * can restore the parent process in the next callback. If the child process executes
+ * execve successfully, parent process gets active again and we hit parent process
+ * restoring it's state
+ *
+ * STEP6:
+ * We free the initial trap and interrupt drakvuf
+ */
+event_response_t handle_execve(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    injector_t injector = (injector_t)info->trap->data;
+
+    event_response_t event;
+
+    switch (injector->step)
+    {
+        case STEP1: // Finds vdso and sets up mmap
+        {
+            memcpy(&injector->saved_regs, info->regs, sizeof(x86_registers_t));
+
+            if (!init_syscalls(drakvuf, info))
+                return override_step(injector, STEP5, VMI_EVENT_RESPONSE_NONE);
+
+            // don't remove the initial trap
+            // it is used for cleanup after restoring registers
+
+            if (!setup_mmap_syscall(injector, info->regs, FILE_BUF_SIZE))
+            {
+                // clear post_syscall_trap
+                free_bp_trap(drakvuf, injector, injector->bp);
+                return override_step(injector, STEP5, VMI_EVENT_RESPONSE_NONE);
+            }
+
+            event = VMI_EVENT_RESPONSE_SET_REGISTERS;
+
+            break;
+        }
+        case STEP2: // forks the process
+        {
+            if (!call_mmap_syscall_cb(injector, info->regs))
+                return cleanup(injector, info->regs);
+
+            char* proc_name = g_malloc0(strlen(info->proc_data.name));
+            strcpy(proc_name, info->proc_data.name);
```

Comment:
```suggestion
            char* proc_name = strdup(info->proc_data.name);
```

Reply:
I also think that you needed to allocalte "strlen(info->proc_data.name) + 1" elements by g_malloc0 as there is '\0' terminating symbol at the end of the string, is that correct or I don't know smth 'bout g_malloc0?

Reply:
And it's also would be nice to use strncpy instead of strcpy :) But yeah, strdup will be better

Reply:
> is that correct or I don't know smth 'bout g_malloc0?

hmm, even I am not sure about it but yes, I think your reasoning makes sense. I did not account for the null in the end.

> or maybe strndup?

ig this won't help us much as `strdup` will allocate the buffer size itself removing the possibility of buffer overflow which might have occurred in `strcpy`